crates/aptos-rest-client/src/lib.rs

```diff
@@ -501,86 +508,192 @@ impl Client {
         &self,
         hash: HashValue,
         expiration_timestamp_secs: u64,
-    ) -> AptosResult<Response<Transaction>> {
-        const DEFAULT_TIMEOUT: Duration = Duration::from_secs(60);
+        server_lag_timeout_after_expiry: Option<Duration>,
+        timeout_from_call: Option<Duration>,
+        fetch: F,
+    ) -> AptosResult<Response<T>>
+    where
+        F: Fn(HashValue) -> Fut,
+        Fut: Future<Output = AptosResult<WaitForTransactionResult<T>>>,
+    {
         const DEFAULT_DELAY: Duration = Duration::from_millis(500);
 
         let start = std::time::Instant::now();
-        while start.elapsed() < DEFAULT_TIMEOUT {
-            let resp = self.get_transaction_by_hash_inner(hash).await?;
-            if resp.status() != StatusCode::NOT_FOUND {
-                let txn_resp: Response<Transaction> = self.json(resp).await?;
-                let (transaction, state) = txn_resp.into_parts();
-
-                if !transaction.is_pending() {
-                    if !transaction.success() {
-                        return Err(anyhow!(
-                            "transaction execution failed: {}",
-                            transaction.vm_status()
-                        ))?;
+        loop {
+            let mut chain_timestamp_usecs = None;
+            match fetch(hash).await? {
+                WaitForTransactionResult::Success(result) => {
+                    return Ok(result);
+                }
+                WaitForTransactionResult::FailedExecution(vm_status) => {
+                    return Err(anyhow!(
+                        "Transaction committed on chain, but failed execution: {}",
+                        vm_status
+                    ))?;
+                }
+                WaitForTransactionResult::Pending(state) => {
+                    if expiration_timestamp_secs <= state.timestamp_usecs / 1_000_000 {
+                        return Err(anyhow!("Transaction expired, it is guaranteed it will not be committed on chain.").into());
+                    }
+                    chain_timestamp_usecs = Some(state.timestamp_usecs);
+                }
+                WaitForTransactionResult::NotFound(error) => {
+                    if let RestError::Api(aptos_error_response) = error {
+                        if let Some(state) = aptos_error_response.state {
+                            if expiration_timestamp_secs <= state.timestamp_usecs / 1_000_000 {
+                                return Err(anyhow!("Transaction expired, it is guaranteed it will not be committed on chain.").into());
+                            }
+                            chain_timestamp_usecs = Some(state.timestamp_usecs);
+                        }
+                    } else {
+                        return Err(error);
                     }
-                    return Ok(Response::new(transaction, state));
+                    sample!(
+                        SampleRate::Duration(Duration::from_secs(30)),
```

Comment:
Nit: This will only ever really trigger once right, do people really make transactions with expiration longer than 30 seconds?

Reply:
in txn-emitter, we set these to much higher. 

but in case someone does want a longer expiry, this makes sense, I think (I don't see that configurable from CLI though)